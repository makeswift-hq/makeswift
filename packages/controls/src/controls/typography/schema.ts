import { z } from 'zod'

import { responsiveValue } from '../../common/schema'

export function typography<C extends z.ZodTypeAny>(color: C) {
  return z.object({
    id: z.string().optional(),
    style: responsiveValue(
      z.object({
        fontFamily: z.string().nullable().optional(),
        lineHeight: z.number().nullable().optional(),
        letterSpacing: z.number().nullable().optional(),
        fontWeight: z.number().nullable().optional(),
        textAlign: z.string().nullable().optional(),
        uppercase: z.boolean().nullable().optional(),
        underline: z.boolean().nullable().optional(),
        strikethrough: z.boolean().nullable().optional(),
        italic: z.boolean().nullable().optional(),
        fontSize: z
          .object({
            value: z.number().nullable(),
            unit: z.string().nullable(),
          })
          .nullable()
          .optional(),
        color: color.optional(),
      }),
    ),
  })
}
