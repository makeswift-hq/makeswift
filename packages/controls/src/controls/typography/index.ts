export * as TypographySchema from './schema'
export * from './typography'
