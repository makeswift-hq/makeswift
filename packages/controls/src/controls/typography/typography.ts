import { z } from 'zod'

import { isNotNil } from '../../lib/functional'
import { safeParse, type ParseResult } from '../../lib/zod'

import { type CopyContext } from '../../context'
import { Targets, type IntrospectionTarget } from '../../introspection'
import {
  type DeserializedRecord,
  type SerializedRecord,
} from '../../serialization'

import { ControlDefinition, serialize } from '../definition'
import { DefaultControlInstance, type SendMessage } from '../instance'

import * as Schema from './schema'

type Schema = typeof Definition.schema
type DataType = z.infer<Schema['data']>
type ValueType = z.infer<Schema['value']>
type ResolvedValueType = z.infer<Schema['resolvedValue']>
type InstanceType = DefaultControlInstance

class Definition extends ControlDefinition<
  typeof Definition.type,
  unknown,
  DataType,
  ValueType,
  ResolvedValueType,
  InstanceType
> {
  static readonly type = 'makeswift::controls::typography' as const

  static get schema() {
    const type = z.literal(Definition.type)

    const data = Schema.typography(
      z
        .object({
          swatchId: z.string().nullable(),
          alpha: z.number().nullable(),
        })
        .nullable(),
    )

    const resolvedData = Schema.typography(
      z
        .object({
          swatch: z
            .object({
              hue: z.number().nullable(),
              saturation: z.number().nullable(),
              lightness: z.number().nullable(),
            })
            .nullable()
            .optional(),
          alpha: z.number().nullable(),
        })
        .optional()
        .nullable(),
    )

    const value = data
    const resolvedValue = z.string()

    const definition = z.object({
      type,
    })

    return {
      type,
      definition,
      data,
      resolvedData,
      value,
      resolvedValue,
    }
  }

  static deserialize(data: DeserializedRecord): Definition {
    if (data.type !== Definition.type) {
      throw new Error(
        `Typography: expected type ${Definition.type}, got ${data.type}`,
      )
    }

    Definition.schema.definition.parse(data)
    return new (class Typography extends Definition {})()
  }

  constructor() {
    super({})
  }

  get controlType() {
    return Definition.type
  }

  get schema() {
    return Definition.schema
  }

  safeParse(data: unknown | undefined): ParseResult<DataType | undefined> {
    return safeParse(this.schema.data, data)
  }

  fromData(data: DataType | undefined): ValueType | undefined {
    return this.schema.data.optional().parse(data)
  }

  toData(value: ValueType): DataType {
    return value
  }

  copyData(
    data: DataType | undefined,
    context: CopyContext,
  ): DataType | undefined {
    if (data == null) return data

    return {
      id:
        context.replacementContext.typographyIds.get(data.id ?? '') ?? data.id,
      style: data.style.map((override) => ({
        ...override,
        value: {
          ...override.value,
          ...(override.value.color == null
            ? {}
            : {
                // Can't use Color().copyData() here as typography's color
                // definition accepts null for swatchId and alpha, while Color
                // Data does not.
                color: {
                  ...override.value.color,
                  alpha: override.value.color.alpha ?? null,
                  swatchId:
                    context.replacementContext.swatchIds.get(
                      override.value.color.swatchId ?? '',
                    ) ??
                    override.value.color.swatchId ??
                    null,
                },
              }),
        },
      })),
    }
  }

  createInstance(sendMessage: SendMessage) {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }

  introspect<R>(data: DataType | undefined, target: IntrospectionTarget<R>) {
    if (data == null) return []

    if (target.type === Targets.Swatch.type) {
      return (
        data.style.flatMap(
          (typography) => target.introspect(typography.value.color) ?? [],
        ) ?? []
      )
    }

    if (target.type === Targets.Typography.type) {
      return [data.id].filter(isNotNil) as R[]
    }

    return []
  }
}

export class unstable_TypographyDefinition extends Definition {}

export function unstable_Typography(): unstable_TypographyDefinition {
  return new unstable_TypographyDefinition()
}
