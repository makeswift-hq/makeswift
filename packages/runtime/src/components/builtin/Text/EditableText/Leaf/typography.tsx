import { CSSObject } from '@emotion/serialize'
import { ReactNode } from 'react'
import { Swatch, Typography as TypographyDTO } from '../../../../../api'
import { DeviceOverride, ResponsiveValue } from '../../../../../prop-controllers'
import { getTypographyStyleSwatchIds } from '../../../../../prop-controllers/introspection'
import { useTypography, useSwatches } from '../../../../../runtimes/react/hooks/makeswift-api'
import { useStyle } from '../../../../../runtimes/react/use-style'
import { colorToString } from '../../../../utils/colorToString'
import { findDeviceOverride, shallowMergeFallbacks } from '../../../../utils/devices'
import { isNonNullable } from '../../../../utils/isNonNullable'
import { responsiveStyle } from '../../../../utils/responsive-style'
import { ColorValue } from '../../../../utils/types'

type TypographyValue = TypographyDTO['style'][number]['value']

export type Typography = {
  id: string | null | undefined
  style: TypographyDTO['style']
}

type EnhancedTypographyValue = {
  fontFamily: string | null
  lineHeight: number | null
  letterSpacing: number | null
  fontWeight: number | null
  textAlign: string | null
  uppercase: boolean | null
  underline: boolean | null
  strikethrough: boolean | null
  italic: boolean | null
  fontSize: { value: number | null; unit: string | null } | null
  color: ColorValue | null
}

export type EnhancedTypography = Array<DeviceOverride<EnhancedTypographyValue>>

export function isDeviceOverride(
  value: TypographyDTO['style'][number],
): value is DeviceOverride<TypographyValue> {
  return value.deviceId === 'desktop' || value.deviceId === 'tablet' || value.deviceId === 'mobile'
}

const withColor =
  (swatches: Swatch[]) =>
  (
    deviceRawTypographyValue: DeviceOverride<TypographyValue>,
  ): DeviceOverride<EnhancedTypographyValue> => {
    const { value, deviceId } = deviceRawTypographyValue

    if (value.color == null) {
      return {
        deviceId,
        value: {
          ...value,
          color: null,
        },
      }
    }
    return {
      deviceId,
      value: {
        ...value,
        color: {
          swatch: swatches.find(s => s && s.id === value.color?.swatchId),
          alpha: value.color?.alpha ?? undefined,
        },
      },
    }
  }

const getDeviceId = ({ deviceId }: DeviceOverride<unknown>) => deviceId

export default function useEnhancedTypography(value: Typography): EnhancedTypography {
  const typography = useTypography(value?.id ?? null)
  const source = typography?.style.filter(isDeviceOverride) ?? []
  const override = value?.style.filter(isDeviceOverride) ?? []

  const swatchIds = [
    ...getTypographyStyleSwatchIds(value?.style),
    ...getTypographyStyleSwatchIds(typography?.style),
  ]
  const swatches = useSwatches(swatchIds).filter(isNonNullable)

  const enhancedSource = source.map(withColor(swatches))
  const enhancedOverride = override.map(withColor(swatches))

  const devices = [
    ...new Set(enhancedSource.map(getDeviceId).concat(enhancedOverride.map(getDeviceId))),
  ]

  return devices
    .map(deviceId => {
      const deviceSource = findDeviceOverride(enhancedSource, deviceId)?.value
      const deviceOverride = findDeviceOverride(enhancedOverride, deviceId)?.value

      if (deviceSource && deviceOverride) {
        return {
          deviceId,
          value: { ...deviceSource, ...deviceOverride },
        }
      } else if (deviceOverride) {
        return {
          deviceId,
          value: deviceOverride,
        }
      } else if (deviceSource) {
        return {
          deviceId,
          value: deviceSource,
        }
      }
      return null
    })
    .filter(isNonNullable)
}

export function useTypographyClassName(value: EnhancedTypography): string {
  return useStyle(
    responsiveStyle<
      EnhancedTypographyValue,
      [ResponsiveValue<EnhancedTypographyValue> | null | undefined]
    >(
      [value],
      ([value]) => {
        if (value === undefined) return {}

        let styles: CSSObject = {}
        if (value.color != null) styles.color = colorToString(value.color)
        if (value.fontFamily != null) styles.fontFamily = value.fontFamily
        if (value.fontSize != null && value.fontSize.value != null && value.fontSize.unit != null)
          styles.fontSize = `${value.fontSize.value}${value.fontSize.unit}`
        if (value.fontWeight != null) styles.fontWeight = value.fontWeight
        if (value.lineHeight != null) styles.lineHeight = value.lineHeight
        if (value.letterSpacing != null) styles.letterSpacing = `${value.letterSpacing / 10}em`
        if (value.uppercase != null)
          styles.textTransform = value.uppercase === true ? 'uppercase' : 'initial'
        if (value.underline != null || value.strikethrough != null)
          styles.textDecoration = [
            Boolean(value.underline) && 'underline',
            Boolean(value.strikethrough) && 'line-through',
          ]
            .filter(Boolean)
            .join(' ')
        if (value.italic != null) styles.fontStyle = value.italic === true ? 'italic' : 'initial'

        return styles
      },
      shallowMergeFallbacks,
    ),
  )
}

export type TypographyLeafProps = {
  typography: Typography
  children: ReactNode
}

export function TypographyLeaf({ typography, children, ...props }: TypographyLeafProps) {
  const enhancedTypography = useEnhancedTypography(typography)
  const typographyClassName = useTypographyClassName(enhancedTypography)
  return (
    <span {...props} className={typographyClassName}>
      {children}
    </span>
  )
}
