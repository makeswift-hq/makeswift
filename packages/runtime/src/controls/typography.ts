export type TypographyControlData = {
  id?: string
  style: Array<{
    deviceId: string
    value: {
      fontFamily?: string | null
      lineHeight?: number | null
      letterSpacing?: number | null
      fontWeight?: number | null
      textAlign?: string | null
      uppercase?: boolean | null
      underline?: boolean | null
      strikethrough?: boolean | null
      italic?: boolean | null
      fontSize?: { value: number | null; unit: string | null } | null
      color?: { swatchId: string | null; alpha: number | null } | null
    }
  }>
} | null

export const TypographyControlType = 'makeswift::controls::typography'

export type TypographyControlDefinition = {
  type: typeof TypographyControlType
}

export function unstable_Typography(): TypographyControlDefinition {
  return {
    type: TypographyControlType,
  }
}
