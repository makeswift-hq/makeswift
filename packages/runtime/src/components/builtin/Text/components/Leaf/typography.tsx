import { CSSObject } from '@emotion/serialize'
import { RenderLeafProps } from 'slate-react'
import { Swatch, Typography } from '../../../../../api'
import { TypographyText } from '../../../../../controls'
import { DeviceOverride, ResponsiveValue } from '../../../../../prop-controllers'
import { getTypographyStyleSwatchIds } from '../../../../../prop-controllers/introspection'
import { useTypography, useSwatches } from '../../../../../runtimes/react/hooks/makeswift-api'
import { useStyle } from '../../../../../runtimes/react/use-style'
import { colorToString } from '../../../../utils/colorToString'
import { findDeviceOverride, shallowMergeFallbacks } from '../../../../utils/devices'
import { isNonNullable } from '../../../../utils/isNonNullable'
import { responsiveStyle } from '../../../../utils/responsive-style'
import { ColorValue } from '../../../../utils/types'

export type TypographyValue = Typography['style'][number]['value']

type EnhancedColor = {
  color?: ColorValue
}

type EnhancedTypographyValue = Omit<TypographyValue, keyof EnhancedColor> & EnhancedColor

export type EnhancedTypography = Array<DeviceOverride<EnhancedTypographyValue>>

export function isDeviceOverride(value: {
  deviceId: string
  value: TypographyValue
}): value is DeviceOverride<TypographyValue> {
  return value.deviceId === 'desktop' || value.deviceId === 'tablet' || value.deviceId === 'mobile'
}

const withColor =
  (swatches: Swatch[]) =>
  (
    deviceRawTypographyValue: DeviceOverride<TypographyValue>,
  ): DeviceOverride<EnhancedTypographyValue> => {
    const { value, deviceId } = deviceRawTypographyValue

    if (value.color == null) {
      const { color, ...nextValue } = value
      return {
        deviceId,
        value: nextValue,
      }
    }
    return {
      deviceId,
      value: {
        ...value,
        color: {
          swatch: swatches.find(s => s && s.id === value.color?.swatchId),
          alpha: value.color?.alpha ?? undefined,
        },
      },
    }
  }

const getDeviceId = ({ deviceId }: DeviceOverride<unknown>) => deviceId

/**
 * `enhanced` here just means typography ids have been replaced with the related entity.
 */
export default function useEnhancedTypography(value: Typography): EnhancedTypography {
  const typography = useTypography(value?.id ?? null)
  const source = typography?.style.filter(isDeviceOverride) ?? []
  const override = value?.style.filter(isDeviceOverride) ?? []

  const swatchIds = [
    ...getTypographyStyleSwatchIds(value?.style),
    ...getTypographyStyleSwatchIds(typography?.style),
  ]
  const swatches = useSwatches(swatchIds).filter(isNonNullable)

  const enhancedSource = source.map(withColor(swatches))
  const enhancedOverride = override.map(withColor(swatches))

  const devices = [
    ...new Set(enhancedSource.map(getDeviceId).concat(enhancedOverride.map(getDeviceId))),
  ]

  return devices
    .map(deviceId => {
      const deviceSource = findDeviceOverride(enhancedSource, deviceId)?.value
      const deviceOverride = findDeviceOverride(enhancedOverride, deviceId)?.value

      if (deviceSource && deviceOverride) {
        return {
          deviceId,
          value: { ...deviceSource, ...deviceOverride },
        }
      } else if (deviceOverride) {
        return {
          deviceId,
          value: deviceOverride,
        }
      } else if (deviceSource) {
        return {
          deviceId,
          value: deviceSource,
        }
      }
      return null
    })
    .filter(isNonNullable)
}

export function useTypographyClassName(value: EnhancedTypography): string {
  return useStyle(
    responsiveStyle<
      EnhancedTypographyValue,
      [ResponsiveValue<EnhancedTypographyValue> | null | undefined]
    >(
      [value],
      ([value]) => {
        if (value === undefined) return {}

        let styles: CSSObject = {}
        if (value.color != null) styles.color = colorToString(value.color)
        if (value.fontFamily != null) styles.fontFamily = value.fontFamily
        if (value.fontSize != null && value.fontSize.value != null && value.fontSize.unit != null)
          styles.fontSize = `${value.fontSize.value}${value.fontSize.unit}`
        if (value.fontWeight != null) styles.fontWeight = value.fontWeight
        if (value.lineHeight != null) styles.lineHeight = value.lineHeight
        if (value.letterSpacing != null) styles.letterSpacing = `${value.letterSpacing / 10}em`
        if (value.uppercase != null)
          styles.textTransform = value.uppercase === true ? 'uppercase' : 'initial'
        if (value.underline != null || value.strikethrough != null)
          styles.textDecoration = [
            Boolean(value.underline) && 'underline',
            Boolean(value.strikethrough) && 'line-through',
          ]
            .filter(Boolean)
            .join(' ')
        if (value.italic != null) styles.fontStyle = value.italic === true ? 'italic' : 'initial'

        return styles
      },
      shallowMergeFallbacks,
    ),
  )
}

export interface TypographyRenderLeafProps extends RenderLeafProps {
  leaf: TypographyText
}

export function TypographyLeaf({ leaf, children, attributes }: TypographyRenderLeafProps) {
  const enhancedTypography = useEnhancedTypography(leaf.typography)
  const typographyClassName = useTypographyClassName(enhancedTypography)
  return (
    <span {...attributes} className={typographyClassName}>
      {children}
    </span>
  )
}
