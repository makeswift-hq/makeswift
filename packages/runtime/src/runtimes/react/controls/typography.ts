import { CSSObject } from '@emotion/serialize'
import { unstable_TypographyDefinition, type DataType } from '@makeswift/controls'

import { useSwatches, useTypography } from '../hooks/makeswift-api'
import { Typography, Swatch } from '../../../api'
import { colorToString } from '../../../components/utils/colorToString'
import { useResponsiveStyle } from '../../../components/utils/responsive-style'
import { ColorValue } from '../../../components/utils/types'
import { DeviceOverride, ResponsiveValue } from '../../../prop-controllers'
import { findBreakpointOverride, shallowMergeFallbacks } from '../../../state/modules/breakpoints'
import { isNonNullable } from '../../../utils/isNonNullable'
import { useStyle } from '../use-style'
import { useBreakpoints } from '../hooks/use-breakpoints'

export function typographyFragementToTypographyControlData(
  typography: Typography | null,
): DataType<unstable_TypographyDefinition> | undefined {
  if (typography == null) return undefined
  return {
    id: typography.id,
    style: typography.style.map(({ deviceId, value }) => ({
      deviceId,
      value: {
        fontFamily: value.fontFamily ?? undefined,
        lineHeight: value.lineHeight ?? undefined,
        letterSpacing: value.letterSpacing ?? undefined,
        fontWeight: value.fontWeight ?? undefined,
        textAlign: value.textAlign ?? undefined,
        uppercase: value.uppercase ?? undefined,
        underline: value.underline ?? undefined,
        strikethrough: value.strikethrough ?? undefined,
        italic: value.italic ?? undefined,
        fontSize: value.fontSize ?? undefined,
        color: value.color ?? undefined,
      },
    })),
  }
}

type EnhancedColor = {
  color?: ColorValue
}

export type TypographyControlDataValue = Exclude<
  DataType<unstable_TypographyDefinition>,
  undefined
>['style'][number]['value']

type EnhancedTypographyValue = Omit<TypographyControlDataValue, keyof EnhancedColor> & EnhancedColor

export type EnhancedTypography = Array<DeviceOverride<EnhancedTypographyValue>>

export function getTypographyStyleSwatchIds(
  style:
    | Exclude<DataType<unstable_TypographyDefinition>, undefined>['style']
    | Typography['style']
    | null
    | undefined,
): string[] {
  return (
    style
      ?.map(override => override.value)
      .flatMap(typographyStyle => typographyStyle.color?.swatchId)
      .filter(isNonNullable) ?? []
  )
}

const withColor =
  (swatches: Swatch[]) =>
  (
    deviceRawTypographyValue: DeviceOverride<TypographyControlDataValue>,
  ): DeviceOverride<EnhancedTypographyValue> => {
    const { value, deviceId } = deviceRawTypographyValue

    if (value.color == null) {
      const { color, ...nextValue } = value
      return {
        deviceId,
        value: nextValue,
      }
    }
    return {
      deviceId,
      value: {
        ...value,
        color: {
          swatch: swatches.find(s => s && s.id === value.color?.swatchId),
          alpha: value.color?.alpha ?? undefined,
        },
      },
    }
  }

const getDeviceId = ({ deviceId }: DeviceOverride<unknown>) => deviceId

/**
 * `enhanced` here just means typography ids have been replaced with the related entity.
 */
export default function useEnhancedTypography(
  value?: DataType<unstable_TypographyDefinition> | null,
): EnhancedTypography {
  const typography = typographyFragementToTypographyControlData(useTypography(value?.id ?? null))
  const source = typography?.style ?? []
  const override = value?.style ?? []
  const breakpoints = useBreakpoints()

  const swatchIds = [
    ...getTypographyStyleSwatchIds(value?.style),
    ...getTypographyStyleSwatchIds(typography?.style),
  ]
  const swatches = useSwatches(swatchIds).filter(isNonNullable)

  const enhancedSource = source.map(withColor(swatches))
  const enhancedOverride = override.map(withColor(swatches))

  const devices = [
    ...new Set(enhancedSource.map(getDeviceId).concat(enhancedOverride.map(getDeviceId))),
  ]

  return devices
    .map(deviceId => {
      const deviceSource = findBreakpointOverride(breakpoints, enhancedSource, deviceId)?.value
      const deviceOverride = findBreakpointOverride(
        breakpoints,
        enhancedOverride,
        deviceId,
        v => v,
      )?.value

      if (deviceSource && deviceOverride) {
        return {
          deviceId,
          value: { ...deviceSource, ...deviceOverride },
        }
      } else if (deviceOverride) {
        return {
          deviceId,
          value: deviceOverride,
        }
      } else if (deviceSource) {
        return {
          deviceId,
          value: deviceSource,
        }
      }
      return null
    })
    .filter(isNonNullable)
}

export function useTypographyClassName(value: EnhancedTypography): string {
  return useStyle(
    useResponsiveStyle<
      EnhancedTypographyValue,
      [ResponsiveValue<EnhancedTypographyValue> | null | undefined]
    >(
      [value],
      ([value]) => {
        if (value === undefined) return {}

        let styles: CSSObject = {}
        if (value.color != null) styles.color = colorToString(value.color)
        if (value.fontFamily != null) styles.fontFamily = value.fontFamily
        if (value.fontSize != null && value.fontSize.value != null && value.fontSize.unit != null)
          styles.fontSize = `${value.fontSize.value}${value.fontSize.unit}`
        if (value.fontWeight != null) styles.fontWeight = value.fontWeight
        if (value.lineHeight != null) styles.lineHeight = value.lineHeight
        if (value.letterSpacing != null) styles.letterSpacing = `${value.letterSpacing / 10}em`
        if (value.uppercase != null)
          styles.textTransform = value.uppercase === true ? 'uppercase' : 'initial'
        if (value.underline != null || value.strikethrough != null)
          styles.textDecoration = [
            Boolean(value.underline) && 'underline',
            Boolean(value.strikethrough) && 'line-through',
          ]
            .filter(Boolean)
            .join(' ')
        if (value.italic != null) styles.fontStyle = value.italic === true ? 'italic' : 'initial'

        return styles
      },
      shallowMergeFallbacks,
    ),
  )
}

export type TypographyControlValue = string

export function useTypographyValue(
  data: DataType<unstable_TypographyDefinition> | undefined,
): TypographyControlValue {
  // for each breakpoint fetch related resources and merge its value with its override
  const enhancedTypography = useEnhancedTypography(data)

  // for each breakpoint shallow merge back up through the breakpoints and create a className
  return useTypographyClassName(enhancedTypography)
}
