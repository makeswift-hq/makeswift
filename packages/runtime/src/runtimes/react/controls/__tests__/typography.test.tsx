import { describe, expect, test } from 'vitest'

import { enhancedTypography } from '../typography'
import * as Fixtures from './fixtures/typography'

describe('GIVEN enhancedTypography', () => {
  test('WHEN hierarchical typography data is passed THEN style overrides are correctly flattened', () => {
    const result = enhancedTypography(
      Fixtures.typographyValue,
      Fixtures.baseTypography,
      Fixtures.breakpoints,
      Fixtures.swatches,
    )

    expect(result).toMatchSnapshot()
  })
})
