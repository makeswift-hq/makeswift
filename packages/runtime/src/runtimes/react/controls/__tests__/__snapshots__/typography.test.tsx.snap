// Vitest Snapshot v1

exports[`GIVEN enhancedTypography > WHEN hierarchical typography data is passed THEN style overrides are correctly flattened 1`] = `
[
  {
    "deviceId": "desktop",
    "value": {
      "color": {
        "alpha": 1,
        "swatch": {
          "__typename": "Swatch",
          "hue": 56,
          "id": "U3dhdGNoOmNlZmExY2E1LTM1NzktNDBiYi1iZjY0LWFkNzY0ZTJkOGY2MQ==",
          "lightness": 53.5,
          "saturation": 100,
        },
      },
      "fontFamily": "Oswald",
      "fontSize": {
        "unit": "px",
        "value": 72,
      },
      "fontWeight": 700,
      "italic": false,
      "letterSpacing": undefined,
      "lineHeight": undefined,
      "strikethrough": undefined,
      "textAlign": undefined,
      "underline": undefined,
      "uppercase": undefined,
    },
  },
  {
    "deviceId": "tablet",
    "value": {
      "color": {
        "alpha": 1,
        "swatch": {
          "__typename": "Swatch",
          "hue": 204,
          "id": "U3dhdGNoOmE0Mzg5YjA5LWQwMWUtNDY5OS1iMGJkLWM2ZDY2NmM1MDVkZQ==",
          "lightness": 50,
          "saturation": 100,
        },
      },
      "fontFamily": "Oswald",
      "fontSize": {
        "unit": "px",
        "value": 72,
      },
      "fontWeight": 700,
      "italic": false,
      "letterSpacing": undefined,
      "lineHeight": undefined,
      "strikethrough": undefined,
      "textAlign": undefined,
      "underline": undefined,
      "uppercase": undefined,
    },
  },
  {
    "deviceId": "mobile",
    "value": {
      "color": {
        "alpha": 1,
        "swatch": {
          "__typename": "Swatch",
          "hue": 204,
          "id": "U3dhdGNoOmE0Mzg5YjA5LWQwMWUtNDY5OS1iMGJkLWM2ZDY2NmM1MDVkZQ==",
          "lightness": 50,
          "saturation": 100,
        },
      },
      "fontFamily": "Oswald",
      "fontSize": {
        "unit": "px",
        "value": 96,
      },
      "fontWeight": 700,
      "italic": false,
      "letterSpacing": undefined,
      "lineHeight": undefined,
      "strikethrough": undefined,
      "textAlign": undefined,
      "underline": undefined,
      "uppercase": undefined,
    },
  },
]
`;
